src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,267 @@
+
+    /**
+     * A pretty print of a code line; debugging.
+     */
+    fun printme() {
+        val df = SimpleDateFormat("yyyy-MM-dd HH:mm z")
+        val fd = df.format(Date(from.commit.getCommitTime().toLong() * 1000))
+        val td = df.format(Date(to.commit.getCommitTime().toLong() * 1000))
+        val fc = "${from.commit.getName()} '${from.commit.getShortMessage()}'"
+        val tc = "${to.commit.getName()} '${to.commit.getShortMessage()}'"
+        println("Line '$text' - '${from.file}:${from.line}' added in $fc $fd")
+        println("  last known as '${to.file}:${to.line}' in $tc $td")
+    }
+}
+
+/**
+ * Used to compute age of code lines in the repo.
+ */
+class CodeLongevity(repoPath: String, tailRev: String) {
+    val repo = FileRepository(repoPath)
+    val head: RevCommit =
+        RevWalk(repo).parseCommit(repo.resolve("refs/heads/master"))
+    val tail: RevCommit? =
+        if (tailRev != "") RevWalk(repo).parseCommit(repo.resolve(tailRev))
+        else null
+
+    /**
+     * A list of all code lines, both alive and deleted, between the given
+     * revisions.
+     */
+    var codeLines: MutableList<CodeLine> = mutableListOf()
+
+    init {
+        compute()
+    }
+
+    // TODO(alex) debugging, remove it
+    fun ohNoDoesItReallyWork(email: String) {
+        var sum: Long = 0
+        var total: Long = 0
+        for (line in codeLines) {
+            val author = line.from.commit.getAuthorIdent()
+            if (author.getEmailAddress() != email) {
+                continue
+            }
+            line.printme()
+            println("  Age: ${line.age} secs")
+            sum += line.age
+            total++
+        }
+
+        //println("All lines:")
+        //codeLines.forEach { line -> line.printme() }
+
+        var avg = if (total > 0) sum / total else 0
+        println("avg code line age for <$email> is ${avg} seconds, lines total: ${total}")
+    }
+
+    /**
+     * Scans through the repo for alive and deleted code lines, and stores them
+     * in the [codeLines] list.
+     */
+    private fun compute() {
+        val treeWalk = TreeWalk(repo)
+        treeWalk.setRecursive(true)
+        treeWalk.addTree(head.getTree())
+
+        val files: MutableMap<String, ArrayList<RevCommitLine>> = mutableMapOf()
+
+        // Build a map of file names and their code lines.
+        while (treeWalk.next()) {
+            val path = treeWalk.getPathString()
+            val lines = getFileLines(treeWalk.getObjectId(0), path, head)
+            if (lines != null) {
+                files.put(path, lines)
+            }
+        }
+  
+        val df = DiffFormatter(DisabledOutputStream.INSTANCE)
+        df.setRepository(repo)
+        df.setDetectRenames(true)
+
+        val revWalk = RevWalk(repo)
+        revWalk.markStart(head)
+
+        var commit: RevCommit? = revWalk.next()  // move the walker to the head
+        while (commit != null && commit != tail) {
```

Comment:
if tail it not empty and not a tail (just a random revision), then we have to crash, no?